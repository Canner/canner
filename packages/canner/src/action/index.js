export { default as actionToMutation } from './actionToMutation';
export { default as actionsToVariables } from './actionsToVariables';
export { default as mutate } from './mutate';
export { generateAction } from './generateAction';
export { ActionManager } from './actionManager';
