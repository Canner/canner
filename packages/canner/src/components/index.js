// @flow

import React, {useRef, forwardRef, useImperativeHandle, useMemo} from 'react';
import {Parser, Traverser} from 'canner-compiler';
import {pickBy} from 'lodash';
import { ApolloProvider } from 'react-apollo';
import Generator from './Generator';
import ListForm from './form/ListForm';
import UpdateForm from './form/UpdateForm';
import CreateForm from './form/CreateForm';
import Page from './form/Page';
// hooks
import useProvider from '../hooks/useProvider';
import useListForm from '../hooks/useListForm';
import useUpdateForm from '../hooks/useUpdateForm';
import useCreateForm from '../hooks/useCreateForm';
import useFormType, {FORM_TYPE} from '../hooks/useFormType';
// i18n
import en from 'react-intl/locale-data/en';
import zh from 'react-intl/locale-data/zh';
import {IntlProvider, addLocaleData} from 'react-intl';
import cannerLocales from './locale';
import componentLocales from '@canner/antd-locales';
addLocaleData([...en, ...zh]);

// type
import type {CMSProps} from './types';

type Props = CMSProps;

export default forwardRef(CannerCMS);

function CannerCMS({
  schema,
  dataDidChange = () => {},
  afterDeploy = () => {},
  baseUrl = '/',
  intl = {
    locale: 'en',
    defaultLocale: 'en',
    messages: {
      'en': {}
    }
  },
  routerParams = {},
  routes = [],
  goTo,
  hideButtons,
  errorHandler = defaultErrorHandler,
  defaultKey,
  client
}: Props, ref) {
  const {visitors, pageSchema, imageStorages, fileStorages, dict} = schema;
  const dataSchema = schema.schema;
  const uiSchema = useMemo(() => ({
    ...pageSchema,
    ...pickBy(dataSchema, v => !v.defOnly)
  }), [pageSchema, dataSchema])
  const componentTree = useMemo(() => compile(uiSchema, visitors), [uiSchema, visitors]);
  const currentLocale = intl.locale || 'en';
  const provider = useProvider({
    schema: dataSchema,
    routes,
    client,
    dataDidChange,
    errorHandler,
    afterDeploy
  });
  const {
    isListForm,
    isUpdateForm,
    isCreateForm,
    isPage
  } = useFormType({routes, routerParams, schema: uiSchema, defaultKey, goTo});
  const listFormProps = useListForm({provider, schema: uiSchema, routes, isListForm});
  const updateFormProps = useUpdateForm({provider, schema: uiSchema, routes, isUpdateForm});
  const createFormProps = useCreateForm({provider, schema: uiSchema, routes, isCreateForm});
  useImperativeHandle(ref, () => ({
    deploy: provider.deploy,
    reset: provider.reset,
  }));
  const commonFormProps = {
    schema: dataSchema,
    baseUrl,
    goTo,
    routes,
    routerParams,
    defaultKey,
    hideButtons,
    componentTree,
    imageStorage: imageStorages[routes[0]],
    fileStorage: fileStorages[routes[0]],
  };
  return (
    <IntlProvider
      locale={currentLocale}
      key={currentLocale}
      defaultLocale={intl.defaultLocale || currentLocale}
      messages={{
        ...(componentLocales[currentLocale] || {}),
        ...(cannerLocales[currentLocale] || {}),
        ...(dict[currentLocale] || {}),
        ...((intl.messages || {})[currentLocale] || {})
      }}
    >
      <ApolloProvider client={client}>
        <React.Fragment>
          {
            isListForm && <ListForm
              {...commonFormProps}
              {...listFormProps}
            >
              <Generator formType={FORM_TYPE.LIST} />
            </ListForm>
          }
          {
            isUpdateForm && <UpdateForm
              {...commonFormProps}
              {...updateFormProps}
            >
              <Generator formType={FORM_TYPE.UPDATE} />
            </UpdateForm>
          }
          {
            isCreateForm && <CreateForm
              {...commonFormProps}
              {...createFormProps}
            >
              <Generator formType={FORM_TYPE.CREATE} />
            </CreateForm>
          }
          {
            isPage && <Page {...commonFormProps}>
              <Generator formType={FORM_TYPE.PAGE} />
            </Page>
          }
        </React.Fragment>
      </ApolloProvider>
    </IntlProvider>
  )
}

function compile(schema, visitors) {
  const parser = new Parser();
  const tree = parser.parse(schema);
  const traverser = new Traverser(tree);
  visitors.forEach(visitor => {
    traverser.addVisitor(visitor);
  });
  const componentTree = traverser.traverse();
  return componentTree;
}

function defaultErrorHandler(e) {
  console.log(e);
  // return notification.error({
  //   message: e.message,
  //   placement: 'bottomRight'
  // });
}
