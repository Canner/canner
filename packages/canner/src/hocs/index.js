import errorCatch from './errorCatch';
import graphqlQuery from './graphqlQuery';
import withCanner from './withCanner';
import withCannerLayout from './withCannerLayout';
import withCache from './withCache';
export default {
  errorCatch,
  graphqlQuery,
  withCanner,
  withCannerLayout,
  withCache
};