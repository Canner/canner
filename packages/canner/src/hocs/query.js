// @flow

import * as React from 'react';
import {List} from 'react-content-loader';
import Toolbar from '../components/toolbar/index';
import {Icon, Spin} from 'antd';
import {mapValues, isNil} from 'lodash';
import {getValue, defaultValue} from './utils';
import {withApollo} from 'react-apollo';
import type {HOCProps} from './types';
type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

@withApollo
export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<HOCProps, State> {
    key: string;
    subscription: any;

    constructor(props: HOCProps) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData().then(() => {
        this.subscribe();
      });
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    componentDidUpdate(prevProps: HOCProps) {
      const {refId} = prevProps;
      if (refId.toString() !== this.props.refId.toString()) {
        // refetch when route change
        this.unsubscribe();
        this.queryData(this.props).then(() => {
          this.subscribe();
        });
      }
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe} = this.props;
      const subscription = subscribe(this.key, result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
      this.subscription = subscription;
    }

    queryData = (props?: HOCProps): Promise<*> => {
      const {fetch} = props || this.props;
      this.setState({
        isFetching: true
      });
      return fetch(this.key).then(result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
    }

    updateData = (data: any, rootValue: any) => {
      const {refId} = this.props;
      const newValue = getValue(data, refId.getPathArr());
      this.setState({
        originRootValue: data,
        rootValue,
        value: newValue,
        isFetching: false
      });
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      const {updateQuery} = this.props;
      updateQuery(paths, args)
        .then(rewatch => {
          if (rewatch) {
            this.unsubscribe();
            this.queryData();
            this.subscribe();
          } else {
            this.queryData();
          }
        });
    }

    render() {
      const {value, isFetching, rootValue, originRootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, pattern, keyName, request, deploy, nullable} = this.props;
      if (!originRootValue) {
        return <List style={{maxWidth: '600px'}}/>;
      }
      if (pattern === 'array') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v.substr(1)]);

        return (
          <Toolbar items={items}
            toolbar={toolbar}
            args={args}
            query={query}
            refId={refId}
            keyName={keyName}
            originRootValue={originRootValue}
            rootValue={rootValue}
            getValue={getValue}
            updateQuery={this.updateQuery}
            request={request}
            deploy={deploy}
          >
            <SpinWrapper isFetching={isFetching}>
              <Com {...this.props} />
            </SpinWrapper>
          </Toolbar>
        );
      } else if (type === 'relation' && relation.type === 'toOne') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={(value && value.id) ? value : defaultValue(type, relation)} />
        );
      } else if (type === 'relation' && relation.type === 'toMany') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue('array', undefined, nullable) : value}/>
        );
      }

      return (
        <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue(type, relation, nullable) : value} />
      );
    }
  };
}

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

function SpinWrapper({
  isFetching,
  children,
  ...props
}: {
  isFetching: boolean,
  children: React.Element<*>
}): React.Element<*> {
  return (
    <Spin indicator={antIcon} spinning={isFetching}>
      {React.cloneElement(children, props)}
    </Spin>
  )
}

export function shouldUpdate(value: any, newValue: any) {
  return value != newValue;
}
