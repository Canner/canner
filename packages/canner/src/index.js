export default from './components';
export {hocs} from './hocs';
export {default as Provder} from './components/Provider';
export {default as ReactRouterProvider} from './components/routeProvider/ReactRouterProvider';
