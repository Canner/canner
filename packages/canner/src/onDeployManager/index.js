// @flow

import {set, unset, get} from 'lodash';

export class OnDeployManager {
  _map: {
    [string]: {
      [string]: Function
    }
  };
  _map = {}
  execute = ({
    key,
    value
  }: {
    key: string,
    value: any
  }): any => {
    const callbacks = this.findCallback(key);
    return callbacks.reduce((result, callback) => callback(result), {data: value});
  }

  findCallback = (key: string): Array<any> => {
    return Object.values(get(this._map, [key], {}));
  }

  registerCallback = (key: string, callback: Function) => {
    const callbackId = randomStr();
    set(this._map, [key, callbackId], callback);
    return callbackId;
  }

  unregisterCallback = (key: string, callbackId: ?string) => {
    unset(this._map, [key, callbackId]);
  }
}

function randomStr() {
  return Math.random().toString(36).substr(2, 6);
}