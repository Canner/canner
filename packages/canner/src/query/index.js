// @flow

import {schemaToQueriesObject, objectToQueries} from './utils';
import {get, set, mapKeys} from 'lodash';

export class Query {
  schema: Object
  queries: Object
  variables: Object

  constructor({schema}: {schema: Object}) {
    this.schema = schema;
    const {queriesObj, variables} = schemaToQueriesObject(schema);
    this.variables = variables;
    this.queries = queriesObj;
  }

  updateQueries = (pathArr: Array<string>, field: string, value: any) => {
    const path = `${pathArr.join('.fields.')}.${field}`;
    if (field === 'args') {
      const args = get(this.queries, path);
      this.variables = {...this.variables, ...Object.keys(args).reduce((result: Object, key: string) => {
        result[args[key]] = value[key];

        return result;
      }, {})};
    } else {
      set(this.queries, path, value);
    }
  }

  getQueries = (pathArr: Array<string>): Object => {
    if (!pathArr || pathArr.length === 0) {
      return this.queries;
    }
    const path = pathArr.join('.fields.');
    return get(this.queries, path);
  }

  toGQL = (key?: string): string => {
    const variables = this.getVairables();
    if (key) {
      const obj = this.getQueries([key]);
      return objectToQueries({[key]: obj}, !obj.declareArgs, variables);  
    } else {
      return objectToQueries(this.queries, false, variables);
    }
  }

  getQueryKey = (key: string): string => {
    // return the string which includes the information of this query excluding the fields and declareArgs
    // e.g: postsConnection(first: $postsFirst,where: $postsWhere)
    const queries = this.getQueries([key]);
    delete queries.fields;
    delete queries.declareArgs;
    return objectToQueries({[key]: queries}, false)
  }

  getVairables = () => {
    return mapKeys(this.variables, (value, key) => key.substr(1));
  }
}