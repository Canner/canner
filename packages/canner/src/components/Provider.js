/**
 * @flow
 */

import React, {useEffect, useState, useRef} from 'react';
import {HOCContext} from '../hocs/context';
import {ApolloProvider} from 'react-apollo';
import pluralize from 'pluralize';
import {actionToMutation, actionsToVariables} from '../action';
import gql from 'graphql-tag';
import {objectToQueries} from '../query/utils';
import {groupBy, difference, mapValues, upperFirst} from 'lodash';
import log from '../utils/log';
import { parseConnectionToNormal } from '../hocs/utils';
import useActionManager from '../hooks/useActionManager';
import useQuery from '../hooks/useQuery';
import useApollo from '../hooks/useApollo';
import useOnDeployManager from '../hooks/useOnDeployManager';
import useCache from '../hooks/useCache';
import type {ProviderProps} from './types';
import type {Action, ActionType} from '../action/types';

type Props = ProviderProps;

export default function Provider({
  schema,
  routes,
  client,
  dataDidChange,
  errorHandler,
  children,
  afterDeploy
}: Props) {
  // ensure these instance only create at first rendering
  const actionManager = useActionManager();
  const query = useQuery(schema);
  const apollo = useApollo({
    schema,
    routes,
    client
  });
  const cache = useCache();
  const onDeployManager = useOnDeployManager();

  const [changedData, setChangedData] = useState(null);
  

  const updateChangedData = () => {
    const actions = actionManager.getActions();
    let changedData = groupBy(actions, (action => action.payload.key));
    changedData = mapValues(changedData, value => {
      if (value[0].type === 'UPDATE_OBJECT') {
        return true;
      }
      return value.map(v => v.payload.id);
    });
    if (dataDidChange) {
      dataDidChange(changedData);
    }
    setChangedData(changedData)
  };

  const fetch = (key: string, refetch: boolean = false): Promise<*> => {
    const queryKey = query.getQueryKey(key);
    // if the data is cached, return it
    if (cache.isCached(queryKey) && !refetch) {
      const cachedData = cache.getData(queryKey);
      log('fetch', 'cached', cachedData);
      return Promise.resolve(cachedData);
    }
    // get the data from apollo and store in cache
    return apollo.fetch(key, query).then(({data}) => {
      const cachedData = {
        data,
        rootValue: parseConnectionToNormal(data)
      }
      log('fetch', 'apollo', cachedData);
      cache.setData(queryKey, cachedData);
      return cachedData;
    });
  };

  const subscribe = (key: string, callback: Function): any => {
    const queryKey = query.getQueryKey(key);
    const subscriptionId = cache.subscribe(queryKey, callback);
    return {
      unsubscribe: () => cache.unsubscribe(queryKey, subscriptionId)
    }
  }

  const updateQuery = (paths: Array<string>, args: Object): Promise<*> => {
    query.updateQueries(paths, 'args', args);
    const variables = query.getVariables();
    log('updateQuery', {variables})
    return fetch(paths[0], true)
      .then(() => true);
  };

  const reset = (key: string = routes[0], id?: string): Promise<*> => {
    if (actionManager.getActions(key, id).length === 0) {
      return Promise.resolve();
    }
    actionManager.removeActions(key, id);
    const queryKey = query.getQueryKey(key);
    cache.removeData(queryKey);
    apollo.reset();
    updateChangedData();

    return Promise.resolve();
  }

  const deploy = (key: string, id?: string): Promise<*> => {
    let actions = actionManager.getActions(key, id);
    if (!actions || !actions.length) {
      return Promise.resolve();
    }
    actions = removeIdInCreateArray(actions);
    const mutation = objectToQueries(actionToMutation(actions[0]), false);
    const variables = actionsToVariables(actions, schema);
    const queryKey = query.getQueryKey(key);
    const cachedData = cache.getData(queryKey);
    const mutatedData = cachedData.data[key];
    const {error} = onDeployManager.publish(key, mutatedData);
    if (error) {
      errorHandler && errorHandler(new Error('Invalid field'));
      return Promise.reject(error);
    }
    return client.mutate({
      mutation: gql`${mutation}`,
      variables
    }).then(result => {
      reset();
      log('deploy', key, {
        id,
        result,
        mutation,
        variables
      });
      actionManager.removeActions(key, id);
      return result.data;
    }).then(result => {
      updateChangedData();
      afterDeploy && afterDeploy({
        key,
        id: id || '',
        result,
        actions
      });
      return result;
    }).catch(e => {
      errorHandler && errorHandler(e);
      log('deploy', e, key, {
        id,
        mutation,
        variables
      });
      // to hocs
      throw e;
    });
  };

  const request = (action: Array<Action<ActionType>> | Action<ActionType>, options: {write: boolean} = {write: true}): Promise<*> => {
    const {write = true} = options;
    const actions = [].concat(action);
    if (actions.length === 0) {
      return Promise.resolve();
    }
    actions.forEach(ac => actionManager.addAction(ac));
    updateChangedData();
    if (write) {
      const data = updateCachedData(actions);
      log('request', action, data);
    }
    return Promise.resolve();
  }

  const updateCachedData = (actions: Array<Action<ActionType>>) => {
    const queryKey = query.getQueryKey(routes[0]);
    cache.mutate(queryKey, actions);
    return cache.getData(queryKey);
  }

  return (
    <ApolloProvider client={client}>
      {/* $FlowFixMe */}
      <HOCContext.Provider value={{
        request: request,
        deploy: deploy,
        fetch: fetch,
        reset: reset,
        updateQuery,
        subscribe,
        query: query,
        onDeploy: onDeployManager.subscribe,
        removeOnDeploy: onDeployManager.unsubscribe,
        dataChanged: changedData
      }}>
        {/* $FlowFixMe */}
        {React.cloneElement(children, {
          request: request,
          deploy: deploy,
          fetch: fetch,
          reset: reset,
          updateQuery,
          subscribe,
          query: query,
          onDeploy: onDeployManager.subscribe,
          removeOnDeploy: onDeployManager.unsubscribe
        })}
      </HOCContext.Provider>
    </ApolloProvider>
  );
}

function removeIdInCreateArray(actions: Array<Action<ActionType>>) {
  return actions.map(action => {
    if (action.type === 'CREATE_ARRAY') {
      const newAction = JSON.parse(JSON.stringify(action));
      delete newAction.payload.value.id;
      delete newAction.payload.value.__typename;
      return newAction;
    }
    return action;
  });
}
