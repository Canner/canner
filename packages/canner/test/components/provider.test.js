import * as React from 'react';
import {render, cleanup, waitForElement} from 'react-testing-library'
import 'jest-dom/extend-expect';
import {schema, defaultData, getAuthorById} from './data';
import Provider from '../../src/components/Provider';
import {createClient, MemoryConnector} from 'canner-graphql-interface';
import {Query} from '../../src/query';
import gql from "graphql-tag";
// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);


function toGQL(schema, key) {
  const query = new Query({schema});
  query.toGQL(key);
  return gql`${query.toGQL(key)}`
}

describe('provider methods', () => {
  let client;
  beforeEach(() => {
    client = createClient({
      schema,
      connector: new MemoryConnector({
        defaultData
      })
    })
  });

  test('fetch object', async () => {
    const {getByTestId} = render(<Provider
      schema={schema}
      client={client}
      routes={[]}
    >
      <div data-testid="children">children</div>
    </Provider>);
    expect(getByTestId('children')).toHaveTextContent('children');
  });
  // TODO: test actions
});
