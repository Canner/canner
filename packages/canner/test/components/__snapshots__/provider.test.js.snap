// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider methods fetch object 1`] = `
Object {
  "home": Object {
    "__typename": "HomePayload",
    "bestAuthor": Object {
      "__typename": "User",
      "age": 10,
      "email": "wwwy3y3@gmail.com",
      "id": "1",
      "images": Array [
        Object {
          "__typename": null,
          "contentType": null,
          "name": null,
          "size": null,
          "url": "url",
        },
      ],
      "name": "user1",
    },
    "count": 10,
    "header": Object {
      "__typename": "homeheader",
      "subTitle": "subTitle",
      "title": "largeTitle",
    },
    "navs": Array [
      Object {
        "__typename": null,
        "text": "nav1",
      },
      Object {
        "__typename": null,
        "text": "nav2",
      },
    ],
    "staredPosts": Array [
      Object {
        "__typename": "Post",
        "id": "1",
        "notes": Array [
          Object {
            "__typename": null,
            "text": "note1",
          },
          Object {
            "__typename": null,
            "text": "note2",
          },
        ],
        "title": "123",
      },
      Object {
        "__typename": "Post",
        "id": "2",
        "notes": Array [
          Object {
            "__typename": null,
            "text": "note3",
          },
          Object {
            "__typename": null,
            "text": "note4",
          },
        ],
        "title": "123",
      },
    ],
  },
}
`;
